stack/clone.go

```diff
@@ -27,62 +28,115 @@ const (
 
 	if !strings.HasPrefix(srcdir, rootdir) {
-		return errors.E(ErrInvalidStackDir, "src dir %q must be inside project root %q", srcdir, rootdir)
+		return 0, errors.E(ErrInvalidStackDir, "src dir %q must be inside project root %q", srcdir, rootdir)
 	}
 
 	if !strings.HasPrefix(destdir, rootdir) {
-		return errors.E(ErrInvalidStackDir, "dest dir %q must be inside project root %q", destdir, rootdir)
+		return 0, errors.E(ErrInvalidStackDir, "dest dir %q must be inside project root %q", destdir, rootdir)
 	}
 
 	if _, err := os.Stat(destdir); err == nil {
-		return errors.E(ErrCloneDestDirExists, destdir)
+		return 0, errors.E(ErrCloneDestDirExists, destdir)
 	}
 
-	srcStack, err := config.LoadStack(root, project.PrjAbsPath(root.HostDir(), srcdir))
-	if err != nil {
-		return errors.E(ErrInvalidStackDir, err, "src dir %q must be a valid stack", srcdir)
+	srcpath := project.PrjAbsPath(rootdir, srcdir)
+
+	// Get all stacks in srcpath (including children)
+	tree, found := root.Lookup(srcpath)
+	if !found {
+		return 0, errors.E(ErrInvalidStackDir, "src dir %q must contain valid stacks", srcdir)
 	}
 
-	logger.Trace().Msg("copying stack files")
+	stackTrees := tree.Stacks()
+	if len(stackTrees) == 0 {
+		return 0, errors.E(ErrInvalidStackDir, "src dir %q must contain valid stacks", srcdir)
+	}
 
-	if err := fs.CopyDir(destdir, srcdir, filterDotFiles); err != nil {
-		return err
+	type cloneTask struct {
+		Srcdir         string
+		Destdir        string
+		ShouldUpdateID bool
 	}
+	tasks := []cloneTask{}
+	srcset := map[string]struct{}{}
+
+	for _, e := range stackTrees {
+		stackSrcdir := e.HostDir()
+		rel, _ := filepath.Rel(srcdir, stackSrcdir)
 
-	if srcStack.ID == "" {
-		logger.Trace().Msg("stack has no ID, nothing else to do")
-		return nil
+		stackDestdir := filepath.Join(destdir, rel)
+		srcset[stackSrcdir] = struct{}{}
+
+		if skipChildStacks && rel != "." {
+			logger.Trace().Msgf("skipping child stack %q", stackSrcdir)
+			continue
+		}
+
+		tasks = append(tasks, cloneTask{
+			Srcdir:         stackSrcdir,
+			Destdir:        stackDestdir,
+			ShouldUpdateID: e.Node.Stack.ID != "",
+		})
 	}
 
-	logger.Trace().Msg("stack has ID, updating ID of the cloned stack")
-	_, err = UpdateStackID(destdir)
-	if err != nil {
-		return err
+	if len(tasks) == 0 {
+		return 0, errors.E(ErrInvalidStackDir, "no stacks to clone in %q", srcdir)
 	}
 
-	return root.LoadSubTree(project.PrjAbsPath(rootdir, destdir))
-}
+	logger.Trace().Msgf("found %v stacks to clone", len(tasks))
+
+	for _, st := range tasks {
+		filter := func(dir string, entry os.DirEntry) bool {
+			if strings.HasPrefix(entry.Name(), ".") {
+				return false
+			}
+
+			abspath := filepath.Join(dir, entry.Name())
+			if _, found := srcset[abspath]; found {
+				return false
+			}
+
+			return true
+		}
+
+		logger.Trace().Msgf("copying stack files from %q to %q", st.Srcdir, st.Destdir)
+
+		if err := fs.CopyDir(st.Destdir, st.Srcdir, filter); err != nil {
```

Comment:
I am wondering about the following edge case:

1. User starts the clone operation
2. The clone operations gets interrupted e.g. due to a file system error, Ctrl+C, crashing process etc.


In that case, would the user have a partially copied directory? 🤔 

I am wondering if we could copy the files to a temporary staging directory first and then execute a "rename" to make this whole process atomic. What do you think?

Reply:
I'm now just deleting the dest dir, since the code is defined right next to the check that it doesn't exist, IMO this is obvious enough.

Reply:
I like @wmalik suggestion but in practice, there are no atomic renames/moves that crosses filesystem boundaries (on linux and other OSs). We implemented like this in vendoring feature and we got bug reports that it didn't work in lots of cases. See here:
https://github.com/terramate-io/terramate/pull/520